core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -808,6 +809,23 @@ public PlanOptimizers(
                         // new join nodes without JoinNode.maySkipOutputDuplicates flag set
                         new OptimizeDuplicateInsensitiveJoins(metadata))));
 
+        // Previous invocations of PushPredicateIntoTableScan do not prune using predicate expression. The invocation in AddExchanges
+        // does this pruning - and we may end up with empty union branches after that. We invoke PushFilterIntoTableScanWithPredicatePruning
+        // here to do this pruning earlier and leverage such removal after that.
+        builder.add(new StatsRecordingPlanOptimizer(
+                optimizerStats,
+                new PushFilterIntoTableScanWithPredicatePruning(plannerContext, typeAnalyzer, statsCalculator)));
+        builder.add(new IterativeOptimizer(
+                plannerContext,
+                ruleStats,
+                statsCalculator,
+                costCalculator,
+                ImmutableSet.of(
+                        new RemoveEmptyUnionBranches(),
+                        new EvaluateEmptyIntersect(),
+                        new RemoveEmptyExceptBranches(),
+                        new TransformFilteringSemiJoinToInnerJoin())));
```

Comment:
Wondering why we need this last rule in here? Is there a case where we see it needed after the union is removed?

Reply:
Not sure If I am missing something but seems it is still used here?